### PR TITLE
RCBC-406 & RCBC-415: Improve conversion of GRPC errors and implement retry handling

### PR DESCRIPTION
**Pull request by DemetrisChr**
* Implemented the retry handling RFC
* Added `Protostellar::ErrorHandling` to convert GRPC errors taking into account the GRPC status's detail block
* All requests are now encapsulated in a `Protostellar::Request` object

### CODE REVIEW

File: lib/couchbase/protostellar/error_handling.rb [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+module Couchbase
+  module Protostellar
+    module ErrorHandling
+      TYPE_URL_PRECONDITION_FAILURE = "type.googleapis.com/google.rpc.PreconditionFailure"
+      TYPE_URL_RESOURCE_INFO = "type.googleapis.com/google.rpc.ResourceInfo"
+      TYPE_URL_ERROR_INFO = "type.googleapis.com/google.rpc.ErrorInfo"
+      TYPE_URL_BAD_REQUEST = "type.googleapis.com/google.rpc.BadRequest"
+
+      def self.handle_grpc_error(grpc_error, request)
```
Comment (DemetrisChr) [inline lib/couchbase/protostellar/error_handling.rb:31]: Added support for a few error types for now, the exact conversions rules have not yet been finalized

File: lib/couchbase/protostellar/error_handling.rb [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        request.context = {} if request.context.nil?
+        detail_block = {}
+
+        rpc_status = grpc_error.to_rpc_status
+        request.context[:grpc_error] = {type: grpc_error.class}
```
Comment (DemetrisChr) [inline lib/couchbase/protostellar/error_handling.rb:36]: Not sure if we want these to be added into the error context, I've added them for now mostly to help with my own debugging.

File: lib/couchbase/protostellar/retry/reason.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        KV_SYNC_WRITE_RE_COMMIT_IN_PROGRESS = Reason.new(true, false).freeze
+        SERVICE_RESPONSE_CODE_INDICATED = Reason.new(true, false).freeze
+        SOCKET_CLOSED_WHILE_IN_FLIGHT = Reason.new(false, false).freeze
+        CIRCUIT_BREAKER_OPEN = Reason.new(true, false).freeze
+        QUERY_PREPARED_STATEMENT_FAILURE = Reason.new(true, false).freeze
+        QUERY_INDEX_NOT_FOUND = Reason.new(true, false).freeze
+        ANALYTICS_TEMPORARY_FAILURE = Reason.new(true, false).freeze
+        SEARCH_TOO_MANY_REQUESTS = Reason.new(true, false).freeze
+        VIEWS_TEMPORARY_FAILURE = Reason.new(true, false).freeze
+        VIEWS_NO_ACTIVE_PARTITION = Reason.new(true, true).freeze
```
Comment (DemetrisChr) [inline lib/couchbase/protostellar/retry/reason.rb:48]: According to the RFC these should be an Enum, I'm not sure what the best way to represent these in Ruby is. The problem with this approach is that after the reasons are added to the request's `retry_attempts` set you can't tell what the reasons were. Should I add a `reason_name` attribute, or add the name of the reason as a Symbol in the set instead, or something else?

File: lib/couchbase/protostellar/client.rb [hunk 4]
```diff
@@ -34,95 +36,43 @@ def initialize(host, credentials, channel_args)
...
+        when :query
+          @query_stub
+        when :routing
+          @routing_stub
+        when :search
+          @search_stub
+        when :view
+          @view_stub
+        else
+          raise ArgumentError "service `#{service}' not recognised"
```
Comment (DemetrisChr) [inline lib/couchbase/protostellar/client.rb:54]: This error is one of those that will not ever be raised unless we did something wrong, should this be a specific type of error?

File: lib/couchbase/protostellar/error_handling.rb [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+          handle_query_error(grpc_error, detail_block, request)
+        else
+          handle_generic_error(grpc_error, detail_block, request)
+        end
+      end
+
+      def self.handle_generic_error(grpc_error, detail_block, request)
+        case grpc_error
+        when GRPC::DeadlineExceeded
+          RequestBehaviour.fail(Couchbase::Error::Timeout.new(nil, request.context))
```
Comment (DemetrisChr) [inline lib/couchbase/protostellar/error_handling.rb:95]: Locally I modified the constructor of the Classic SDK's errors to be able to take the context as an optional argument - so it should not affect the behaviour of the Classic SDK (I think). Should I push that change to the classic SDK and keep this as it is here, or use `instance_variable_set` instead?

**Review by avsej — APPROVED** [review]

**Pull request merged**